File: mirea/cources/2026-2027/aii-july/03/Part-04-Hashing-and-pitfalls/examples/example-02.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

void print_dict(const string &name, const auto &mp) {
    cout << name << " = {";
    bool first = true;
    for (const auto &[key, value] : mp) {
        if (!first) cout << ", ";
        cout << key << ": " << value;
        first = false;
    }
    cout << "}" << '\n';
}

int main() {
    vector<string> words = {"pear", "apple", "kiwi", "pear", "plum", "apple"};

    // Считаем частоты слов.
    unordered_map<string, int> cnt;
    for (const string &word : words) {
        cnt[word]++;
    }

    // Количество различных ключей.
    cout << cnt.size() << '\n';

    print_dict("cnt", cnt);

    // Поиск по ключу.
    for (const string &word : vector<string>{"pear", "mango"}) {
        auto it = cnt.find(word);
        if (it == cnt.end()) {
            cout << "Find " << word << ": fail" << '\n';
        } else {
            cout << "Find " << word << ": ok, value = " << it->second << '\n';
        }
    }

    // Удаление ключей.
    for (const string &word : vector<string>{"pear", "mango", "apple"}) {
        cout << "Delete " << word << ": ";
        auto it = cnt.find(word);
        if (it == cnt.end()) {
            cout << "fail";
        } else {
            cout << "ok";
            cnt.erase(it);
        }
        print_dict(", cnt", cnt);
    }
}

File: mirea/cources/2026-2027/aii-july/03/Part-04-Hashing-and-pitfalls/examples/example-02.py
from collections import defaultdict


def print_dict(name, mp):
    parts = [f"{key}: {value}" for key, value in mp.items()]
    print(f"{name} = {{{', '.join(parts)}}}")


words = ["pear", "apple", "kiwi", "pear", "plum", "apple"]

# Считаем частоты слов.
cnt = defaultdict(int)
for word in words:
    cnt[word] += 1

# Количество различных ключей.
print(len(cnt))

print_dict("cnt", cnt)

# Поиск по ключу.
for word in ["pear", "mango"]:
    if word in cnt:
        print(f"Find {word}: ok, value = {cnt[word]}")
    else:
        print(f"Find {word}: fail")

# Удаление ключей.
for word in ["pear", "mango", "apple"]:
    print(f"Delete {word}: ", end="")
    if word in cnt:
        print("ok", end="")
        del cnt[word]
    else:
        print("fail", end="")
    print_dict(", cnt", cnt)
